apm-sniffer/apm-sdk-plugin/finagle-6.25.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/finagle/CompatibilityChecker.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.apache.skywalking.apm.plugin.finagle;
+
+import org.apache.skywalking.apm.agent.core.logging.api.ILog;
+import org.apache.skywalking.apm.agent.core.logging.api.LogManager;
+
+import static org.apache.skywalking.apm.plugin.finagle.ContextHolderFactory.getLocalContextHolder;
+import static org.apache.skywalking.apm.plugin.finagle.ContextHolderFactory.getMarshalledContextHolder;
+
+public class CompatibilityChecker {
```

Comment:
Version compatibility feature is provided by a thing called `witness` class. If you provide that in the instrumentation definition, your plugin is activated only those class exists.

Reply:
The ThreadLocal has low performance in the agent, and cost more CPU resources, comparing to other ways. Your codes use this in high frequency which is a potential performance issue. We only accept one way to use thread local, is there is no way to communicate between the interceptor. From your class name, this is not for that case.

Reply:
> Version compatibility feature is provided by a thing called `witness` class. If you provide that in the instrumentation definition, your plugin is activated only those class exists.

The implementation detail of this plugin depend on a private class of Finagle Framework, we can not ensure that class will still exists in future versions. The mechanism of witness class can ensure this plugin can work in existing versions, it can not ensure this plugin will still work in future versions. So the purpose of the class  is to check whether this plugin is compatible with future versions, if it does't, the plugin just do nothing, avoiding unexpected runtime exceptions.
This has been added in comments.

Reply:
> The ThreadLocal has low performance in the agent, and cost more CPU resources, comparing to other ways. Your codes use this in high frequency which is a potential performance issue. We only accept one way to use thread local, is there is no way to communicate between the interceptor. From your class name, this is not for that case.

`ContextHolderFactory`?  There are no `ThreadLocal` used directly, however, it use `com.twitter.util.Local` which backed by `ThreadLocal`, but this is necessary, please check the comments.

Reply:
First, That thing liked I said, having IP issue. And same as I said, you shouldn't implement in this way.

I think I have said clear. Any question?

Reply:
The future versions are never being sure. So it is meaningless. That is why we have explicit version test CI process, and indicate them in support list doc. Only real test could verify that.